### PR TITLE
migration vexflow 4.0

### PR DESCRIPTION
**Pull request by rvilarl**
I decide to re-start the migration.

### CODE REVIEW

**Comment by rvilarl** [comment]
Simon I could also update all the packages. There is only one topic that I did not manage to resolve: Reading `process.env. ...` variables

**Comment by sschmidTU** [comment]
If package updates help with part of the migration process, sure, otherwise i would prefer making package updates a separate issue/PR. No need to complicate things.

**Comment by rvilarl** [comment]
@sschmidTU ready to review

**Comment by rvilarl** [comment]
@ronyeh ready to review

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetDrawer.ts [hunk 1]
```diff
@@ -53,15 +54,15 @@ export class VexFlowMusicSheetDrawer extends MusicSheetDrawer {
     public drawSheet(graphicalMusicSheet: GraphicalMusicSheet): void {
         // vexflow 3.x: change default font
         if (this.rules.DefaultVexFlowNoteFont === "gonville") {
-            (Vex.Flow as any).DEFAULT_FONT_STACK = [(Vex.Flow as any).Fonts?.Gonville, (Vex.Flow as any).Fonts?.Bravura, (Vex.Flow as any).Fonts?.Custom];
+            // rvilarl Vex.Flow.setMusicFont("Gonville","Bravura","Custom");
```
Comment (ronyeh) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetDrawer.ts:57]: Why is this commented out? Is it not working?

**Comment by sschmidTU** [comment]
Awesome, thank you Rodrigo! I'll review asap.
Heads up, beforehand i'll release (at least) one more OSMD version with the old Vexflow. I'd like the Vexflow update to be the single change of one version update.
Also, i'll run the visual regression tests, which will be very interesting between the versions. Also, performance tests. (timing the regression tests)

**Comment by ronyeh** [comment]
Exciting! You could merge these changes into a OSMD-VexFlow-4 branch, and then after Mohit publishes a beta, you can update the package-json to point to the vexflow 4 beta. Once VexFlow 4 is officially released, merge the OSMD-VexFlow-4 branch into OSMD main.

**Comment by ronyeh** [comment]
You can also look at the migration for Music21j: https://github.com/cuthbertLab/music21j/pull/144

There, we eliminated use of the Vex.Flow.* prefixes, opting to import the classes directly.

If you have name collisions, you can alias the imports as VFStave, VFTextBracket, etc.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetDrawer.ts (hunk 1 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetDrawer.ts]: Done I just forgot to uncomment.

**Comment by rvilarl** [comment]
@sschmidTU please look at the last commit and you will see what I resolved with the update of the packages. In particular the error:
TypeError: s.StaveNote is not a constructor

Let me know when you have see it and I will remove the last commit. Node 16 did not resolve the problem.

**Comment by rvilarl** [comment]
Ready to review

**Comment by rvilarl** [comment]
@sschmidTU I have intorduced a bug with the cursors, please have a look at the demo.

**Comment by rvilarl** [comment]
@sschmidTU most probably we will get a vexflow 4.0 beta, would you make a beta in your repo as well to let the people start paying with 4.0? This can be on another branch (no master or develop)

**Comment by sschmidTU** [comment]
By the way, some of the `VexFlowPatch` files will still have to be adapted, because they were modifications of the old javascript versions of files in Vexflow. Right now, our npm `prebuild` script will probably have no effect because it copies javascript files to vexflow that aren't used anymore.
Some of the changes in `VexFlowPatch` were already merged to Vexflow, so they can just be removed, but others aren't merged yet.
The readme in `/VexFlowPatch` should give some info.

@rvilarl I'll check the bugs and see if we can make a beta when i have the time.

**Comment by rvilarl** [comment]
@sschmidTU I will upload two images of the bugs so that you can have a quick look

File: src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalNote.ts [hunk 2]
```diff
@@ -76,19 +76,16 @@ export class VexFlowGraphicalNote extends GraphicalNote {
...
-        if (!vfnote) {
-            vfnote = (this.vfnote[0] as any);
-        }
-        const noteheads: any = vfnote.note_heads;
+    public notehead(vfNote?: StemmableNote): {line: number} {
+        const vfnote: StemmableNote = vfNote || this.vfnote[0];
+        const noteheads: NoteHead[] = (vfnote as StaveNote).noteHeads;
         if (noteheads && noteheads.length > this.vfnoteIndex && noteheads[this.vfnoteIndex]) {
-            return vfnote.note_heads[this.vfnoteIndex];
+            return { line: (vfnote as StaveNote).noteHeads[this.vfnoteIndex].getLine() };
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalNote.ts:88]: @sschmidTU this is doing something different, makes sense?

Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalNote.ts:88]: looks fine for now, as long as we only need the line from the Vexflow NoteHead, which is all we're using for now in VexFlowMeasure. I guess the old version doesn't work anymore because Vexflow now returns NoteHead instead of any :)

File: src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalSymbolFactory.ts [hunk 3]
```diff
@@ -88,10 +88,7 @@ export class VexFlowGraphicalSymbolFactory implements IGraphicalSymbolFactory {
      * @returns {VexFlowMeasure}
      */
     public createExtraGraphicalMeasure(staffLine: StaffLine): GraphicalMeasure {
-        const extraGraphicalMeasure: GraphicalMeasure = new VexFlowMeasure(staffLine.ParentStaff, undefined, staffLine);
-        extraGraphicalMeasure.IsExtraGraphicalMeasure = true; // this also means that MeasureNumber < 0 because unchanged
-        extraGraphicalMeasure.ExtraGraphicalMeasurePreviousMeasure = staffLine.Measures.last();
-        return extraGraphicalMeasure;
+        return new VexFlowMeasure(staffLine.ParentStaff, undefined, staffLine);
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalSymbolFactory.ts:91]: I will revert this change, I do not see the point.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts [hunk 4]
```diff
@@ -1242,21 +1248,18 @@ export class VexFlowMeasure extends GraphicalMeasure {
...
                     }
                 }
 
                 // add fingering
                 if (voiceEntry.parentVoiceEntry && this.rules.RenderFingerings) {
-                    if (this.rules.FingeringPosition === PlacementEnum.Left ||
-                        this.rules.FingeringPosition === PlacementEnum.Right) {
-                            this.createFingerings(voiceEntry);
-                    } // else created in MusicSheetCalculater.createFingerings() as Labels
+                    this.createFingerings(voiceEntry);
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:1262]: Revert this change, I do not see the point.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts [hunk 5]
```diff
@@ -1383,15 +1386,6 @@ export class VexFlowMeasure extends GraphicalMeasure {
...
             let fingeringPosition: PlacementEnum = this.rules.FingeringPosition;
-            //currently only relevant for grace notes, because we create other fingerings above/below in MusicSheetCalculator.createFingerings
-            if (this.rules.FingeringPositionGrace === PlacementEnum.AboveOrBelow) {
-                //if (this.rules.FingeringPosition === PlacementEnum.AboveOrBelow) {
-                if (this.isUpperStaffOfInstrument()) { // (e.g. piano right hand)
-                    fingeringPosition = PlacementEnum.Above;
-                } else if (this.isLowerStaffOfInstrument()) {
-                    fingeringPosition = PlacementEnum.Below;
-                }
-            }
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:1394]: Revert this change, I do not see the point.

Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts:1394]: Yes. In general, you shouldn't change OSMD's logic in this PR, only how it interfaces with Vexflow.

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts [hunk 6]
```diff
@@ -937,9 +941,6 @@ export class VexFlowMusicSheetCalculator extends MusicSheetCalculator {
             }
 
             let nextShiftLastMeasure: GraphicalMeasure = nextShiftStaffline.Measures[nextShiftStaffline.Measures.length - 1];
-            if (nextShiftLastMeasure.IsExtraGraphicalMeasure) { // key/rhythm change measure
-              nextShiftLastMeasure = nextShiftStaffline.Measures[nextShiftStaffline.Measures.length - 2];
-            }
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:942]: Revert

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts [hunk 7]
```diff
@@ -949,13 +950,6 @@ export class VexFlowMusicSheetCalculator extends MusicSheetCalculator {
               lastNote = endStaffEntry;
             }
 
-            const logPrefix: string = "VexFlowMusicSheetCalculator.calculateSingleOctaveShift: ";
-            if (!firstNote) {
-              log.warn(logPrefix + "no firstNote found");
-            }
-            if (!lastNote) {
-              log.warn(logPrefix + "no lastNote found");
-            }
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:958]: Revert

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts [hunk 8]
```diff
@@ -976,10 +970,7 @@ export class VexFlowMusicSheetCalculator extends MusicSheetCalculator {
 
   private calculateOctaveShiftSkyBottomLine(startStaffEntry: GraphicalStaffEntry, endStaffEntry: GraphicalStaffEntry,
                                             vfOctaveShift: VexFlowOctaveShift, parentStaffline: StaffLine): void {
-    if (!endStaffEntry) {
-      log.warn("octaveshift: no endStaffEntry");
-      return;
-    }
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:982]: Revert

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts [hunk 9]
```diff
@@ -1002,22 +993,22 @@ export class VexFlowMusicSheetCalculator extends MusicSheetCalculator {
     const startX: number = startStaffEntry.PositionAndShape.AbsolutePosition.x - startXOffset;
     const stopX: number = endStaffEntry.PositionAndShape.AbsolutePosition.x + endXOffset;
     vfOctaveShift.PositionAndShape.Size.width = startX - stopX;
-    const textBracket: Vex.Flow.TextBracket = vfOctaveShift.getTextBracket();
-    const fontSize: number = (textBracket as any).font.size / 10;
+    const textBracket: TextBracket = vfOctaveShift.getTextBracket();
+    const fontSize: number = textBracket.fontSizeInPoints / 10;
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:997]: @sschmidTU fontSizeInPoints or fontSizeInPixels?

Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:997]: pixels. One unit in OSMD is 10 pixels in Vexflow. Not sure how points are handled in Vexflow.

Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts:997]: If a point in Vexflow is 10 pixels for the default font and settings, maybe that's ok.

File: src/MusicalScore/Graphical/VexFlow/VexFlowVoiceEntry.ts [hunk 10]
```diff
@@ -9,32 +10,32 @@ import { ColoringModes } from "./../DrawingParameters";
...
     public applyBordersFromVexflow(): void {
-        const staveNote: any = (this.vfStaveNote as any);
-        if (!staveNote.getNoteHeadBeginX) {
+        const staveNote: StemmableNote = this.vfStaveNote;
+        if (!(staveNote instanceof Vex.Flow.StaveNote)) {
             return;
         }
-        const boundingBox: any = staveNote.getBoundingBox();
-        const modifierWidth: number = staveNote.getNoteHeadBeginX() - boundingBox.x;
+        const boundingBox: BoundingBox = staveNote.getBoundingBox();
```
Comment (rvilarl) [inline src/MusicalScore/Graphical/VexFlow/VexFlowVoiceEntry.ts:24]: import as VFBoundingBox

Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowVoiceEntry.ts:24]: Yes. There might be more naming conflicts with Vexflow, so prefix the Vexflow versions with VF.

Comment (sschmidTU) [reply src/MusicalScore/Graphical/VexFlow/VexFlowVoiceEntry.ts]: It might actually be nice to have the VF prefix for (almost?) every Vexflow class, where we had Vex.Flow.X before.

File: src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalNote.ts (hunk 2 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalNote.ts:88]: ok

File: src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalSymbolFactory.ts (hunk 3 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowGraphicalSymbolFactory.ts]: Done

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts (hunk 4 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts]: Done

File: src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts (hunk 5 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMeasure.ts]: Done

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts (hunk 7 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts]: Done

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts (hunk 8 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts]: Done

File: src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts (hunk 9 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowMusicSheetCalculator.ts]: done

File: src/MusicalScore/Graphical/VexFlow/VexFlowVoiceEntry.ts (hunk 10 above)
Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowVoiceEntry.ts]: I have done it on all the classes with collision. This one was missing.

**Comment by rvilarl** [comment]
Still messing ties and cursor :((

Captura de pantalla de 2021-11-13 18-47-35

**Comment by rvilarl** [comment]
Already rebased to 1.3.0!
@sschmidTU What would you propose to debug the cursor and tie?

**Comment by sschmidTU** [comment]
@rvilarl Not sure, probably just a debugging session and looking at the values and where they come from, and how they are different with the old vexflow. maybe set up two instances of the project, one old, one new, and run them in parallel if possible.
For debugging the cursor, debug into `Cursor.update()` and see where the `x` value comes from.

**Comment by sschmidTU** [comment]
@rvilarl Not sure, probably just a debugging session and looking at the values and where they come from, and how they are different with the old vexflow. maybe set up two instances of the project, one old, one new, and run them in parallel if possible.
For debugging the cursor, debug into Cursor.update() and see where the x value comes from.

See our debugging guide here:
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/wiki/Debugging-(VSCode)

File: src/MusicalScore/Graphical/VexFlow/VexFlowSlur.ts [hunk 11]
```diff
@@ -66,7 +66,7 @@ export class VexFlowSlur {
     public createVexFlowCurve(): void {
             this.vfCurve = new Vex.Flow.Curve( this.vfStartNote,
                                                this.vfEndNote,
-                                               this.curve_Options()
+                                               undefined//this.curve_Options()
```
Comment (sschmidTU) [inline src/MusicalScore/Graphical/VexFlow/VexFlowSlur.ts]: why were the options removed? I assume this is WIP / debugging?

Comment (rvilarl) [reply src/MusicalScore/Graphical/VexFlow/VexFlowSlur.ts]: This is already removed in `develop`, I think that the change comes from the rebase.
